src/main/java/net/snowflake/ingest/streaming/internal/MemoryInfoProviderFromRuntime.java

```diff
@@ -4,20 +4,51 @@
 
 package net.snowflake.ingest.streaming.internal;
 
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
 /** Reads memory information from JVM runtime */
 public class MemoryInfoProviderFromRuntime implements MemoryInfoProvider {
-  @Override
-  public long getMaxMemory() {
-    return Runtime.getRuntime().maxMemory();
+  private final long maxMemory;
+  private volatile long totalFreeMemory;
+  private final ScheduledExecutorService executorService;
+  private static final long FREE_MEMORY_UPDATE_INTERVAL_MS = 1000;
+  private static final MemoryInfoProviderFromRuntime INSTANCE =
+      new MemoryInfoProviderFromRuntime(FREE_MEMORY_UPDATE_INTERVAL_MS);
+
+  private MemoryInfoProviderFromRuntime(long freeMemoryUpdateIntervalMs) {
+    maxMemory = Runtime.getRuntime().maxMemory();
+    totalFreeMemory =
+        Runtime.getRuntime().freeMemory() + (maxMemory - Runtime.getRuntime().totalMemory());
+    executorService =
+        new ScheduledThreadPoolExecutor(
+            1,
+            r -> {
+              Thread th = new Thread(r, "MemoryInfoProviderFromRuntime");
+              th.setDaemon(true);
+              return th;
+            });
+    executorService.scheduleAtFixedRate(
```

Comment:
I'm not sure updating the memory usage every second is enough, the SDK is memory intensive and we have run into many OOM issues before adding this check. We basically want to stop accepting new rows completely ASAP. I feel like we can ask DeltaStream to call insertRows instead of insertRow to fix this issue, WDYT?

Reply:
My gut feeling tells me that this will introduce some regression, and then we will need to tweak the parameter default values, but let's see :)

Reply:
I'm wondering what's the common practice for this? sync vs async? We would need to do similar things in Buffering Tier as well, cc: @sfc-gh-japatel

Reply:
It is much easier in C++. There is no garbage collection and you can track exactly how much memory is being used by keeping rowset buffers in memory without having to ask the runtime.

Reply:
I don't see any problem with the thread continuing to run - it is fairly cheap at this frequency. Even with per-client singleton, if they don't close the client this and other threads would keep running anyways. I'll merge this but if you feel strongly about this, let's chat and I can do a follow-up.